evervault-enclaves/src/main/java/com/evervault/sdk/enclaves/AttestationDocCache.kt

```diff
@@ -1,76 +1,103 @@
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
 
 @Serializable
 data class AttestationDoc(
     @SerialName("attestation_doc") val attestationDoc: String,
 )
 @OptIn(DelicateCoroutinesApi::class)
-class AttestationDocCache(private val enclaveName: String, private val appUuid: String) {
+class AttestationDocCache(
+    private val enclaveName: String,
+    private val appUuid: String,
+    private val enclaveHostname: String? = "enclave.evervault.com",
+) {
     private var attestationDoc: ByteArray = ByteArray(0)
     private val lock = ReentrantReadWriteLock()
+    private val maxAttempts = 3
+    private val initialDelayMs = 1000L
+    private var pollingJob: Job? = null
+    private val client: OkHttpClient = OkHttpClient.Builder()
+        .callTimeout(10000, TimeUnit.MILLISECONDS)
+        .connectTimeout(10000, TimeUnit.MILLISECONDS)
+        .readTimeout(10000, TimeUnit.MILLISECONDS)
+        .build()
 
-    init {
-        GlobalScope.launch(Dispatchers.IO) {
-            storeDoc(2)
-            poll(300)
-        }
-    }
+    private suspend fun storeDocWithBackoff() {
+        var currentAttempt = 0
 
-    private fun storeDoc(retries: Int) {
-        if(retries >= 0) {
+        while (currentAttempt < maxAttempts) {
             try {
-                val url =
-                    "https://${enclaveName}.${appUuid}.enclave.evervault.com/.well-known/attestation"
+                val url = "https://${enclaveName}.${appUuid}.${enclaveHostname}/.well-known/attestation"
                 val response = getDocFromEnclave(url)
                 val decodedDoc = Base64.decode(response.attestationDoc, Base64.DEFAULT)
                 set(decodedDoc)
+                return
             } catch (e: Exception) {
-                Log.w(TAG, "Failed to get attestationdoc ${e.message}")
-                storeDoc(retries - 1)
+                currentAttempt++
+                if (currentAttempt == maxAttempts) {
+                    return
+                }
+
+                val delayMs = initialDelayMs * (1 shl (currentAttempt - 1))
+                Log.d("AttestationDocCache", "Populating doc from cached failed retry count: $currentAttempt, delay $delayMs")
+                Log.d("AttestationDocCache", "Exception getting attestation doc", e)
+                delay(delayMs)
+            }
+        }
+    }
+
+    suspend fun initialize() {
+        if (attestationDoc.isEmpty()) {
+            withContext(Dispatchers.IO) {
+                storeDocWithBackoff()
+            }
+
+            pollingJob = pollingJob ?: GlobalScope.launch(Dispatchers.IO) {
+                poll(300)
             }
         }
     }
 
     fun get(): ByteArray {
         return lock.read {
             if(attestationDoc.isEmpty()) {
-                storeDoc(2)
+                throw IllegalStateException("Attestation doc not initialized. Call initialize() first")
```

Comment:
is this potentially a breaking change? this exception would not have thrown previously.

Reply:
Yes, removed and replaced with a check for the cache in the `AttestationTrustManagerGA` which will cause an `SSLHandshakeException` if it is empty.